Add a breadcrumb trail to pages

JackSekoranja opened the pull request:
These changes will add several clickable links to the top of the page that the user can click to navigate. This is specified easily within the any HTML file extending base.html by typing something similar to the following:
```
{% block breadcrumb-trail %}
    {% breadcrumb  '/page_link' 'Breadcrumb Title' %}
    {% endcrumb 'Non-Clickable Title' %}
{% endblock %}
```

I couldn't figure out how to easily link page urls without explicitly stating the endpoint. It is technically possible using `{% url list:index %}` within pure HTML, but not within Django template blocks (to my knowledge).

![Home](https://user-images.githubusercontent.com/36946090/56896849-4bdc9580-6ad0-11e9-807d-c24d40eeda74.jpg)
![Create Subplan](https://user-images.githubusercontent.com/36946090/56896851-4bdc9580-6ad0-11e9-9920-0737715846e8.jpg)

JackSekoranja committed:
Added a breadcrumb trail to each page on the website

JackSekoranja committed:
Merge branch 'develop' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-breadcrumb_trail

# Conflicts:
#	cassdegrees/templates/managecourses.html
#	cassdegrees/templates/manageprograms.html
#	cassdegrees/templates/managesubplans.html

pep8speaks commented:
Hello @JackSekoranja! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cassdegrees/ui/templatetags/breadcrumbs.py`](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py):

> [Line 6:1](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 8:40](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L8): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 8:51](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L8): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 8:56](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L8): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 8:61](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L8): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 10:1](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L10): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 12:31](https://github.com/cass-degrees/CASS-Degrees-Code/blob/16e057eaff0a7dbaea1a4cbb54b76310517c260b/cassdegrees/ui/templatetags/breadcrumbs.py#L12): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

JackSekoranja committed:
Add breadcrumbs to pages that didn't have them

aidengeorge97 reviewed (approved):
Looks good, I tested it on each page and it works fine. Sensible structure.

Daniel-Jang committed:
Fixed some pep 8 violations - no change in functionality

JackSekoranja committed:
Rename `endcrumb` to `finalcrumb`
This was done since PyCharm thinks any tags of the form `endX` are closing tags. Therefore, this change will remove the warnings on the affected lines

Daniel-Jang commented:
I had a look and I noticed that whenever I click on either Create Program Template or Create Subplan from the home page, this appears:
![image](https://user-images.githubusercontent.com/37033052/56934073-018a0180-6b2e-11e9-8ae3-2a3b6928b948.png)

That link leads to list.html, which wasn't used in this particular case to get to this page (from home -> Create Program Template).

Is this intentional? I wasn't sure if this was a design choice.

Apart from this, it looks good.

JackSekoranja committed:
Merge remote-tracking branch 'origin/feature-breadcrumb_trail' into feature-breadcrumb_trail

# Conflicts:
#	cassdegrees/ui/templatetags/breadcrumbs.py

j-selby commented:
@Daniel-Jang I believe the angle here is that the home page is shortcuts, not direct components

j-selby commented on cassdegrees/templates/createcourse.html:
```diff
@@ -1,9 +1,20 @@
 {% block subtitle %}{% if edit %}Edit{% else %}Create{% endif %} Course{% endblock %}
 
+{% block breadcrumb-trail %}
+    {% breadcrumb  '/list/?view=Course' 'Template/Subplan/Course' %}
```
`Templates/...`
IMO this should be plural

JackSekoranja replied:
Actually this was done deliberately because I figured it made sense in terms to go `Program >> Add` instead of `Programs >> Add`. Happy to change it if you really wanted to but I wasn't sure which was better

JackSekoranja replied:
I decided to add your changes after consultation with Daniel

j-selby reviewed (changes requested):
Looking good!

JackSekoranja commented:
@Daniel-Jang Yeah, it was kinda unavoidable since the breadcrumbs are generated on a per-page basis. I thought it was acceptable since there are `create new X` on the list page anyway. Plus if they want to get to the hompepage they can just click the `home` icon

Daniel-Jang commented:
@j-selby Ah ok, that makes sense, thanks for clearing it up for me.

JackSekoranja committed:
Made `Template/Subplan/Course` plural in all pages
